ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -95,6 +101,29 @@ public void onError(Throwable error) {
         actionListener.onFailure(new OpenSearchStatusException(errorMessage, status));
     }
 
+    private void handleThrottlingInHeader(SdkHttpFullResponse sdkResponse) {
+        if (MapUtils.isEmpty(sdkResponse.headers())) {
+            return;
+        }
+        List<String> errorsInHeader = sdkResponse.headers().get(AMZ_ERROR_HEADER);
+        if (errorsInHeader == null || errorsInHeader.isEmpty()) {
+            return;
+        }
+        boolean containsThrottlingException = errorsInHeader.stream().anyMatch(str -> str.startsWith("ThrottlingException"));
+        if (containsThrottlingException && executionContext.getExceptionHolder().get() == null) {
+            log.error("Remote server returned error code: {}", statusCode);
+            executionContext
+                .getExceptionHolder()
+                .compareAndSet(
```

Comment:
The line 146 will set the exception as http body `{"message":null}`. Have you confirmed this line will happen after line 146 ? If not, the error will still be `{"message":null}`, right ? Have you done some manual testing if IT is hard/flaky to add?

Reply:
Actually it will not anymore because the executionContext.getExceptionHolder() isn't null anymore. Line 146 only sets the exception when there is nothing in the ExceptionHolder. But for the throttling errors, this PR has set an exception in the ExceptionHolder so it will be not be set anymore (line 146 will not invoke). I have verified it using a multi-node cluster.

Reply:
sounds good if onHeaders execute first

Reply:
Verified with multiple test with reasonable TPS concurrency, the onHeader() always executes before onStream(). So when the exception is caught at the headers, it will not be override in the onStream() when receiving the request body.